xde/src/xde.rs

```diff
@@ -2048,16 +2043,17 @@ fn new_port(
 #[unsafe(no_mangle)]
 unsafe extern "C" fn xde_rx(
     arg: *mut c_void,
-    mrh: *mut mac::mac_resource_handle,
     mp_chain: *mut mblk_t,
-    _is_loopback: boolean_t,
-) {
+    out_mp_tail: *mut *mut mblk_t,
+    out_count: *mut c_uint,
+    out_len: *mut usize,
+) -> *mut mblk_t {
     __dtrace_probe_rx(mp_chain as uintptr_t);
 
     // Safety: This arg comes from `Arc::from_ptr()` on the `MacClientHandle`
-    // corresponding to the underlay port we're receiving on. Being
-    // here in the callback means the `MacPromiscHandle` hasn't been
-    // dropped yet and thus our `MacClientHandle` is also still valid.
+    // corresponding to the underlay port we're receiving on (derived from
+    // `DlsStream`). Being here in the callback means the `MacSiphon` hasn't
+    // been dropped yet, and thus our `MacClientHandle` is also still valid.
     let stream: Arc<DlsStream> = unsafe {
         let mch_ptr = arg as *const DlsStream;
         Arc::increment_strong_count(mch_ptr);
```

Comment:
Since we're here:  What's the value in turning this into an `Arc`, if we already have guarantees that the `mch_ptr` will point to the resource covered by the underlying `Arc` which is guaranteed to be "live" through the execution of this function?  With such a guarantee, I think we could just cast the pointer into a reference and skip the increment/decrement of the refcount?